docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,104 @@
+# ADR 024: Coin Metadata
+
+## Changelog
+
+- 05/19/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Assets in the Cosmos SDK are represented via a `Coins` type that consists of an `amount` and a `denom`,
+where the `amount` can be any arbitrarily large or small value. In addition, the Cosmos SDK uses an
+account-based model where there are two types of primary accounts -- basic accounts and module accounts.
+All account types have a set of balances that are composed of `Coins`. The `x/bank` module keeps
+track of all balances for all accounts and also keeps track of the total supply of balances in an
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may chose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+As a result, we can define the type as follows:
+
+```protobuf
+message DenomUnit {
+  string denom    = 1;
+  uint32 exponent = 2;  
+}
+
+message Metadata {
+  string description = 1;
+  string base = 2;
+  repeated DenomUnit denom_units = 3;
+  repeated string aliases = 4;
+}
+```
+
+As an example, the ATOM's metadata can be defined as follows:
+
+```json
+{
+  "description": "The native staking token of the Cosmos Hub.",
+  "base": "uatom",
+  "denom_units": [
+    {
+      "denom": "matom",
+      "exponent": 3,
+    },
+    {
+      "denom": "atom",
+      "exponent": 6,
+    }
+  ],
+  "aliases": [
+    "btcuatom",
+    "kavauatom"
```

Comment:
little confused by these examples, i know these are made up but is it meant to imply that some people call atoms `kavauatom` or that when an atom appears on the kava chain it is referred to as `kavauatom`?

Reply:
I'm confused with the format as well. I assume this is a relayed token? Which format is it using (eg: `sourcechain|denom`)?

Reply:
Relay? Not sure I follow. I think the examples are confusing people, so I'll revise them. 

See the original intent from @sunnya97: https://github.com/cosmos/cosmos-sdk/pull/2072/files#diff-0e8ffa8fe366671daf24c98523830814R23